torch/quantization/convert_modules.py

```diff
@@ -69,7 +113,9 @@ def prepare(module, qconfig_dict):
         qconfig_dict: dictionary that maps from name of submodule to quantization
                       configuration
     """
-    return add_observer(module, qconfig_dict)
+    propogate_qconfig(module, qconfig_dict)
+    add_observer(module)
+    return add_quant_dequant(module)
```

Comment:
We need to add two additional options in qconfig_dict per module:  quantize_input, dequantize_output. 
add_quant_dequant would look at the qconfig of each module and decide whether to insert quant/dequant.
This complicates the qconfigDict specification: For example, if I want to quantize the whole model I would need to say:

qConfigDict={' ': { weight: weight_qconfig, activation:activation_qconfig, quantize_input:True, dequantize_output: True}}
For each of the child modules, assuming all of them are quantized, I would want to specify quantize_input=False and dequantize_output = False. 

One way to solve this issue is to have default for quantize_input = False and dequantize_output = False. If user wants to override it, then qConfigDict needs to provided explicitly.

Reply:
I don't feel we need to do that, user can modify their model directly if they want to have the whole_module functionality. they can skip the first step of inserting quant_dequant and start with add_observers

Reply:
adding this to config is complicates things a lot and does buy much convenience

Reply:
We will have a hierarchy mismatch in this case:
For example, lets have:

`class TwoLayerLinearModel(torch.nn.Module):
    def __init__(self):
        super(TwoLayerLinearModel, self).__init__()
        self.fc1 = torch.nn.Linear(5, 8).to(dtype=torch.float)
        self.fc2 = torch.nn.Linear(8, 5).to(dtype=torch.float)

    def forward(self, x):
        x = self.fc1(x)
        x = self.fc2(x)
        return x


myModel = TwoLayerLinearModel()
tq.prepare(myModel)
print(myModel)
# Would print:
fc1
 Quant
 module
 DeQuant
fc2
 Quant
 module
 DeQuant

#If the user does this manually, by inserting quant/dequant: we would have:
Quant
fc1
fc2
DeQuant
'


self.fc1.module, otherwise, it will show up as: self.fc1.

Reply:
we can ask user to use QuantWrapper to insert quant dequant as well?

Reply:
QuantWrapper might not cover all the cases, in case of more complicated topologies I think it is OK to have different hierarchy since our tool will not work in that case which means there is no corresponding quantized module using our tool.

Reply:
I concur with what Jerry said - we shouldn't overcomplicate this tooling.

As for hierarchy mismatch - **it's ok, you're modifying your model**. Note that state_dict won't cause problems as noone would be loading checkpoints from float version directly.

As for qconfig - since it's recursive it'd work just fine - `fc1` would apply also for `fc1.module`

Reply:
I agree, had a chat with Jerry to discuss this further. Lets keep it simple.  insert quant dequant will only insert quant/dequant for every module, no need to have other functionality.

Reply:
> I concur with what Jerry said - we shouldn't overcomplicate this tooling.
> 
> As for hierarchy mismatch - **it's ok, you're modifying your model**. Note that state_dict won't cause problems as noone would be loading checkpoints from float version directly.
> 
> As for qconfig - since it's recursive it'd work just fine - `fc1` would apply also for `fc1.module`

For QAT, we would be loading state_dict as we train and save checkpoints, but still changing the hierarchy is ok, so I agree we should keep the API simple